sdks/python/apache_beam/transforms/dofn_lifecycle_test.py

```diff
@@ -0,0 +1,91 @@
+    self._setup_called = False
+    self._start_bundle_calls = 0
+    self._finish_bundle_calls = 0
+    self._teardown_called = False
+
+  def setup(self):
+    # Use assert instead of unittest.TestCase.assert* methods because
+    # this is not a TestCase
+    assert not self._setup_called,'setup should not be called twice'
+    assert self._start_bundle_calls == 0, 'setup should be called before start_bundle'
+    assert self._finish_bundle_calls == 0, 'setup should be called before finish_bundle'
+    assert not self._teardown_called, 'setup should be called before teardown'
+    self._setup_called = True
+
+  def start_bundle(self):
+    # Use assert instead of unittest.TestCase.assert* methods because
+    # this is not a TestCase
+    assert self._setup_called, 'setup should have been called'
+    assert self._start_bundle_calls == self._finish_bundle_calls , \
+      'there should be as many start_bundle calls as finish_bundle calls'
+    assert not self._teardown_called, 'teardown should not have been called'
+    self._start_bundle_calls += 1
+
+  def process(self, element):
+    # Use assert instead of unittest.TestCase.assert* methods because
+    # this is not a TestCase
+    assert self._setup_called, 'setup should have been called'
+    assert self._start_bundle_calls > 0, 'start_bundle should have been called'
+    assert self._start_bundle_calls == self._finish_bundle_calls + 1, \
+      'there should be one start_bundle call with no call to finish_bundle'
+    assert not self._teardown_called, 'teardown should not have been called'
+    return [element * element]
+
+  def finish_bundle(self):
+    # Use assert instead of unittest.TestCase.assert* methods because
+    # this is not a TestCase
+    assert self._setup_called, 'setup should have been called'
+    assert self._start_bundle_calls > 0, 'start_bundle should have been called'
+    assert self._start_bundle_calls == self._finish_bundle_calls + 1, \
+      'there should be one start_bundle call with no call to finish_bundle'
+    assert not self._teardown_called, 'teardown should not have been called'
+    self._finish_bundle_calls += 1
+
+  def teardown(self):
+    # Use assert instead of unittest.TestCase.assert* methods because
+    # this is not a TestCase
+    assert self._setup_called, 'setup should have been called'
+    assert self._start_bundle_calls == self._finish_bundle_calls , \
+      'there should be as many start_bundle calls as finish_bundle calls'
+    assert not self._teardown_called,'teardown should not be called twice'
+    # Unfortunately there is no way to assert that teardown was actually called.
+    # This is acceptable because teardown is best-effort.
+    self._teardown_called = True
+
+
+@attr('ValidatesRunner')
+class DoFnLifecycleTest(unittest.TestCase):
+  def test_dofn_lifecycle(self):
+    with TestPipeline() as p:
+      (p
+        | 'Start' >> beam.Create([1, 2, 3])
+        | 'Do' >> beam.ParDo(CallSequenceEnforcingDoFn()))
```

Comment:
How about, adding an assert after the end of the the with: to assert that teardown was actually called.

Reply:
Removed comments.

Reply:
Would that actually work? My understanding is that with remote runners, the DoFn instance here and the one on the worker are different physical instances, and there isn't a way to check for state changes on the DoFn on the worker to figure out if teardown was called or not.

Also, when I was doing print debugging, it looks like teardown is not being called on the direct runner.

Reply:
It would not work with remote runners but it would work DirectRunner. Is not direct runner default for TestPipeline?

> Also, when I was doing print debugging, it looks like teardown is not being called on the direct runner.

That looks like a bug in this PR.

Reply:
I see. If we only run this test for DirectRunner, then I can add the extra assert.

Yes, I will investigate and see if DirectRunner can be made to call teardown.

Reply:
Print debugging shows that teardown is being correctly called. However, the assert in this location fails because the worker is using a different unpickled copy of the DoFn.

- A pickle round trip happens in [ptransform.py](https://github.com/apache/beam/blob/9c8a8dc241296ad5003124c8bcb94a50bb53beb5/sdks/python/apache_beam/transforms/ptransform.py#L689) during pipeline construction time.
- The default direct runner FnApiRunner unpickles the DoFn in [operations.py](https://github.com/apache/beam/blob/9c8a8dc241296ad5003124c8bcb94a50bb53beb5/sdks/python/apache_beam/runners/worker/operations.py#L508) and [bundle_processor.py](https://github.com/apache/beam/blob/9c8a8dc241296ad5003124c8bcb94a50bb53beb5/sdks/python/apache_beam/runners/worker/bundle_processor.py#L996).
- The alternative direct runner BundleBasedDirectRunner performs a pickle round trip in [transform_evaluator.py](https://github.com/apache/beam/blob/9c8a8dc241296ad5003124c8bcb94a50bb53beb5/sdks/python/apache_beam/runners/direct/transform_evaluator.py#L592-L593) if _perform_dofn_pickle_test is true.

This means that it is difficult to get a handle to the DoFn instance that the direct runner worker is actually using.

Reply:
Added a test that checks that a global flag is set. This assumes that the worker is running in the same process as the test.